concent_api/core/message_handlers.py

```diff
@@ -1312,27 +1317,42 @@ def handle_send_subtask_results_verify(
             reason=message.concents.ServiceRefused.REASON.InvalidRequest,
         )
 
-    (requestor_has_enough_deposit, provider_has_enough_deposit) = bankster.claim_deposit(
-        concent_use_case=ConcentUseCase.ADDITIONAL_VERIFICATION,
-        requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
-        provider_ethereum_address=task_to_compute.provider_ethereum_address,
-        subtask_cost=task_to_compute.price,
-    )
-
-    if not requestor_has_enough_deposit:
-        return message.concents.ServiceRefused(
-            reason=message.concents.ServiceRefused.REASON.TooSmallRequestorDeposit,
-        )
-    if not provider_has_enough_deposit:
-        return message.concents.ServiceRefused(
-            reason=message.concents.ServiceRefused.REASON.TooSmallProviderDeposit,
-        )
-
     with transaction.atomic(using='control'):
         subtask = get_one_or_none(
             Subtask.objects.select_for_update(),
             subtask_id=compute_task_def['subtask_id'],
         )
+        if subtask is not None and subtask.state_enum in [
+            Subtask.SubtaskState.VERIFICATION_FILE_TRANSFER,
+            Subtask.SubtaskState.ADDITIONAL_VERIFICATION,
+        ]:
+            return message.concents.ServiceRefused(
+                reason=message.concents.ServiceRefused.REASON.DuplicateRequest,
+            )
+
+        (claim_against_requestor, claim_against_provider) = bankster.claim_deposit(
+            subtask_id=task_to_compute.subtask_id,
+            concent_use_case=ConcentUseCase.ADDITIONAL_VERIFICATION,
+            requestor_ethereum_address=task_to_compute.requestor_ethereum_address,
```

Comment:
This operation can't be performed inside an `atomic` block because it has to be able to commit its transactions (if a transaction already exists, all its commits become a part of the parent transaction).

Reply:
:+1: done, I guess this should work the same in `handle_send_force_subtask_results`? I have changed it there too.

Reply:
Yeah, kinda. But `handle_send_force_subtask_results()` has problems too. It does not nest transactions but it does not put a lock on the subtask. This means that multiple instances could potentially create claims against the same subtask.

I think it's good enough as a workaround for now but we should have an issue for a correct solution. This is a race condition.

Reply:
Unfortunately it's not completely race-condition-free. There's still a small window between transactions in which to processes can interfere. Consider this:
1. First process runs the view, subtask is in a passive state.
2. Second process runs the view, subtask is in a passive state.
3. First process creates claims.
4. Second process creates claims.
5. First process locks the subtask.
6. Second process tries to lock the subtask and waits.
7. First process changes subtask state and commits.
8. Second pocess can now continue. It also sets subtask state and commits.

We end up with two sets of claims for the same subtask. It would be fine the subtask was locked from the beginning but unfortunately it isn't.

As I said, this is probably not something that will happen often so this flow is fine as a temporary workaround but still, it's broken and should be reported so that we can do it properly if we have time later.